AnnoDesigner/Helper/RoadSearchHelper.cs

```diff
@@ -0,0 +1,113 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using AnnoDesigner.Core.Models;
+
+namespace AnnoDesigner.Helper
+{
+    public static class RoadSearchHelper
+    {
+        private static void DoNothing(AnnoObject objectInRange) { }
+
+        public static Dictionary<int, Dictionary<int, AnnoObject>> PrepareGridDictionary(IEnumerable<AnnoObject> placedObjects)
+        {
+            var dictionary = new Dictionary<int, Dictionary<int, AnnoObject>>();
+
+            foreach (var placedObject in placedObjects)
+            {
+                var x = (int)placedObject.Position.X;
+                var y = (int)placedObject.Position.Y;
+                for (var i = 0; i < placedObject.Size.Width; i++)
+                {
+                    if (!dictionary.ContainsKey(x + i))
+                    {
+                        dictionary.Add(x + i, new Dictionary<int, AnnoObject>());
+                    }
+
+                    for (var j = 0; j < placedObject.Size.Height; j++)
+                    {
+                        dictionary[x + i][y + j] = placedObject;
+                    }
+                }
+            }
+
+            return dictionary;
+        }
+
+        public static HashSet<(int x, int y)> BreathFirstSearch(
+            IEnumerable<AnnoObject> placedObjects,
+            IEnumerable<AnnoObject> startObjects,
+            Func<AnnoObject, int> rangeGetter,
+            Action<AnnoObject> inRangeAction = null,
+            Dictionary<int, Dictionary<int, AnnoObject>> gridDictionary = null)
+        {
+            inRangeAction = inRangeAction ?? DoNothing;
+            gridDictionary = gridDictionary ?? PrepareGridDictionary(placedObjects);
+            var blackCells = new HashSet<(int x, int y)>();
+
+            startObjects = startObjects.Where(o => rangeGetter(o) > 0.5).ToList();
+            if (startObjects.Count() == 0)
+                return blackCells;
+
+            var searchedCells = new Queue<(int remainingDistance, int x, int y)>();
+            var blackObjects = new HashSet<AnnoObject>();
```

Comment:
nit: what is it that you mean by `blackCells`/`blackObjects`? Maybe these variables could be better named?

Reply:
Renamed to visitedCells/visitedObjects